sycl/unittests/scheduler/StreamInitDependencyOnHost.cpp

```diff
@@ -0,0 +1,122 @@
+//==------ LinkedAllocaDependencies.cpp --- Scheduler unit tests -----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "SchedulerTest.hpp"
+#include "SchedulerTestUtils.hpp"
+
+#include <detail/scheduler/scheduler_helpers.hpp>
+
+using namespace cl::sycl;
+
+class MockHandler : public sycl::handler {
+public:
+  MockHandler(shared_ptr_class<detail::queue_impl> Queue, bool IsHost)
+      : sycl::handler(Queue, IsHost) {}
+
+  void setType(detail::CG::CGTYPE Type) {
+    static_cast<sycl::handler *>(this)->MCGType = Type;
+  }
+
+  void addStream(const detail::StreamImplPtr &Stream) {
+    sycl::handler::addStream(Stream);
+  }
+
+  unique_ptr_class<detail::CG> finalize() {
+    auto CGH = static_cast<sycl::handler *>(this);
+    unique_ptr_class<detail::CG> CommandGroup;
+    switch (CGH->MCGType) {
+    case detail::CG::KERNEL:
+    case detail::CG::RUN_ON_HOST_INTEL: {
+      CommandGroup.reset(new detail::CGExecKernel(
+          std::move(CGH->MNDRDesc), std::move(CGH->MHostKernel),
+          std::move(CGH->MKernel), std::move(CGH->MArgsStorage),
+          std::move(CGH->MAccStorage), std::move(CGH->MSharedPtrStorage),
+          std::move(CGH->MRequirements), std::move(CGH->MEvents),
+          std::move(CGH->MArgs), std::move(CGH->MKernelName),
+          std::move(CGH->MOSModuleHandle), std::move(CGH->MStreamStorage),
+          CGH->MCGType, CGH->MCodeLoc));
+      break;
+    }
+    default:
+      throw runtime_error("Unhandled type of command group",
+                          PI_INVALID_OPERATION);
+    }
+
+    return CommandGroup;
+  }
+};
+
+using CmdTypeTy = cl::sycl::detail::Command::CommandType;
+
+static bool ValidateDepCommandsTree(const detail::Command *Cmd,
+                                    std::queue<CmdTypeTy> DepCmdsTypes,
+                                    const detail::SYCLMemObjI *MemObj) {
```

Comment:
NIT: That's nice solution for small recursions. Though, when large ones will happen one should pass queue by reference, not by copy. Anyway, I believe this isn't a case here.

Reply:
It a queue is passed by reference then there will be error in this function if there are for example two empty_task dependencies. I'll rewrite it to remove queue copying.

Reply:
Updated function.